File: graph_theory/warshall_floyd.cpp
#include <bits/stdc++.h>
#define INF 100000000
#define rep(i,n) for(int i=0; i<n; ++i)
using namespace std;

typedef vector<vector<int>> M; M dp;

int warshall_floyd(int n) {
    rep(k,n) {
        rep(i,n) {
            rep(j,n){
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j]);
            }
        }
    }
    return 0;
}

int main() {
    int n, w; cin >> n >> w;
    dp = M(n, vector<int>(n, INF));
    rep(i,n) {
        rep(j,n){
            dp[i][j] = INF;
        }
    }
    rep(i,n) dp[i][i] = 0;
    rep(i,w) {
        int x, y, z;
        cin >> x >> y >> z;
        dp[x][y] = z;
        // dp[y][x] = z; // 有向グラフなら消す
    }
    warshall_floyd(n);
    rep(i, n) {
        rep(j, n) {
            if (i != j && dp[i][j] != INF)
            cout << "cost from " << i << " to " << j << ": " << dp[i][j] << endl;
        }
    }
    return 0;
}

File: graph_theory/warshall_floyd.py
def warshall_floyd(d, n):
    # d[i][j] = iからjへの最短距離
    for k in range(n):
        for i in range(n):
            for j in range(n):
                d[i][j] = min(d[i][j], d[i][k] + d[k][j])

    return d

def main():
    n, w = map(int, input().split())
    d = [[float('inf') for _ in range(n)] for _ in range(n)]
    for i in range(w):
        x, y, z = map(int, input().split())
        d[x][y] = z
        # d[y][x] = z #有向グラフなら消す
    for i in range(n):
        d[i][i] = 0
    print(warshall_floyd(d, n))

if __name__ == '__main__':
    main()


"""
sample input:

7 10
0 1 2
0 2 5
1 2 4
2 3 2
1 3 6
1 4 10
3 5 1
4 5 3
4 6 5
5 6 9

expected output:
[0, 2, 5, 7, 11, 8, 16]
[2, 0, 4, 6, 10, 7, 15]
[5, 4, 0, 2, 6, 3, 11]
[7, 6, 2, 0, 4, 1, 9]
[11, 10, 6, 4, 0, 3, 5]
[8, 7, 3, 1, 3, 0, 8]
[16, 15, 11, 9, 5, 8, 0]

"""
